From 671b3ff32efec7d42fb46fd342ff9793df595e51 Mon Sep 17 00:00:00 2001
From: zhengq <qi.zheng@intel.com>
Date: Fri, 24 Apr 2015 23:49:11 +0800
Subject: [PATCH] Pinctrl: set GPIO mode if calling gpio_request.

Sometimes the initial pad mode values for pins which are
used as GPIO are not correctly set on BIOS. To fix this
issue, set the pad mode as GPIO every time when request
GPIO in the kernel driver.

Change-Id: I07ccfd874a7970f717249ee5e1268d1baa5f949f
Tracked-On: https://jira01.devtools.intel.com/browse/AKP-1476
Signed-off-by: zhengq <qi.zheng@intel.com>
---
 drivers/pinctrl/pinctrl-cherryview.c | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/drivers/pinctrl/pinctrl-cherryview.c b/drivers/pinctrl/pinctrl-cherryview.c
index b695ef6..4f1f214 100644
--- a/drivers/pinctrl/pinctrl-cherryview.c
+++ b/drivers/pinctrl/pinctrl-cherryview.c
@@ -320,8 +320,8 @@ struct gpio_pad_info {
 	/* Interrupt line selected (0~15), -1 if not interruptible. */
 	int interrupt_line;
 
-	/*function mode switch support. Once enable, gpio_request will force set
-	* the pad mode to GPIO, gpio_free will switch it back to native mode.
+	/*function mode switch support. Once enable, gpio_free will
+	* switch it back to native mode.
 	* It is disabled in default. */
 	int switch_en;
 
@@ -780,13 +780,9 @@ static int chv_gpio_request(struct gpio_chip *chip, unsigned offset)
 
 	dev_dbg(&cg->pdev->dev, "chv_gpio_request %d\n", offset);
 
-	if (cg->pad_info[offset].switch_en > 0) {
-		/*change it to GPIO mode*/
-		value = chv_readl(reg) | CV_GPIO_EN;
-		chv_writel(value, reg);
-		dev_info(&cg->pdev->dev,
-			"pin %d force set to gpio mode\n", offset);
-	}
+	/*make sure it is configured to GPIO mode*/
+	value = chv_readl(reg) | CV_GPIO_EN;
+	chv_writel(value, reg);
 
 	return 0;
 }
-- 
1.9.1

